test/e2e/test/apmserver/checks_apm.go

```diff
@@ -119,15 +126,125 @@ func (c *apmClusterChecks) CheckApmServerVersion(apm apmv1.ApmServer) test.Step
 	}
 }
 
+// CheckIndexCreation ensure that, prior to attempting to ingest events, the APM Server user
+// has the necessary permissions to either create indexes (ES < 8.x), or index documents into
+// non-existing indexes (ES >= 8.x). This fixes a transient issue that happens when upgrading
+// Elasticsearch between major versions where it takes a bit of time to transition between
+// file-based user roles, and permissions errors were being returned by Elasticsearch.
+func (c *apmClusterChecks) CheckIndexCreation(apm apmv1.ApmServer, k *test.K8sClient) test.Step {
```

Comment:
I'm a bit confused by this test, how do we know that creating documents in an index named `"apm-testindex-"` or `metrics-apm.testindex-` is enough to validate that we can later make a call to the `/intake/v2/events` API?

Wouldn't it be easier to try calling the `/intake/v2/events` API a first time, until it succeeds, to check that we can use that API ? Maybe by reusing the `CheckEventsAPI` in a retry loop.

Reply:
>  it's the underlying Elasticsearch permissions moving from v7 -> v8 APM file-based roles where the real issue is

Just to be sure I understand, could we use the `user/_privileges` API (`curl -k -u e2e-mercury-apmserver-upgrade-XXX-apm-user:REDACTED https://XXXXXX:9200/_security/user/_privileges`) and compare the result with the [predefined roles](https://github.com/elastic/cloud-on-k8s/blob/e15d93fc76178aa6d0801e4d3193dbf2ffd79277/pkg/controller/elasticsearch/user/roles.go#L78-L99) ?

With the 7.x credentials:

```json
{
	"cluster": ["cluster:admin/xpack/monitoring/bulk", "manage_api_key", "manage_ilm", "monitor"],
	"global": [],
	"indices": [{
		"names": ["apm-*"],
		"privileges": ["create_doc", "create_index", "manage"],
		"allow_restricted_indices": false
	}, {
		"names": [".monitoring-beats-*"],
		"privileges": ["create_doc", "create_index"],
		"allow_restricted_indices": false
	}],
	"applications": [],
	"run_as": []
}
```

With the 8x permissions:

```json
{
	"cluster": ["cluster:admin/xpack/monitoring/bulk", "cluster:monitor/main", "manage_index_templates", "monitor"],
	"global": [],
	"indices": [{
		"names": ["traces-apm.sampled-*"],
		"privileges": ["maintenance", "monitor", "read"],
		"allow_restricted_indices": false
	}, {
		"names": [".monitoring-beats-*"],
		"privileges": ["create_doc", "create_index"],
		"allow_restricted_indices": false
	}, {
		"names": ["logs-apm*", "metrics-apm*", "traces-apm*"],
		"privileges": ["auto_configure", "create_doc"],
		"allow_restricted_indices": false
	}],
	"applications": [],
	"run_as": []
}
```

Once the `user/_privileges` returns the expected result we could then proceed with the next step ?

Reply:
I was thinking along similar lines here https://github.com/elastic/cloud-on-k8s/pull/5470#discussion_r839629287

Reply:
@barkbay as @pebrc mentioned, I did some testing down that path, and we certainly could go with that approach if we feel that it's a cleaner, or more clear approach, but the implementation for that approach is going to be more involved that simply testing to see if an index can be created, or document indexed.  I'll also do some testing shortly on just wrapping `test.Eventually` on the "send message -> check for document" steps, and maybe that will be cleaner than all the other approaches to this... I'll update today with some results.

Reply:
@barkbay @pebrc e2e test updated to just leverage `test.Eventually`, and it's certainly more straightforward.

Reply:
Test from `7.17.2` to `8.2.0-SNAPSHOT` failed on 8c0a532b5:

```
=== RUN   TestAPMServerVersionUpgradeToLatest8x/ApmServer_endpoint_should_eventually_be_reachable#01
Retries (5m0s timeout): .
    --- PASS: TestAPMServerVersionUpgradeToLatest8x/ApmServer_endpoint_should_eventually_be_reachable#01 (1.63s)
=== RUN   TestAPMServerVersionUpgradeToLatest8x/ApmServer_version_should_be_the_expected_one#01
    --- PASS: TestAPMServerVersionUpgradeToLatest8x/ApmServer_version_should_be_the_expected_one#01 (0.06s)
=== RUN   TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01
Retries (5m0s timeout): .
    step.go:43: 
        	Error Trace:	utils.go:88
        	Error:      	Received unexpected error:
        	            	timeout reached after 5m0s
        	Test:       	TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01
{"log.level":"error","@timestamp":"2022-04-15T16:01:45.582+0200","message":"stopping early","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","error":"test failure","error.stack_trace":"github.com/elastic/cloud-on-k8s/test/e2e/test.StepList.RunSequential\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/step.go:44\ngithub.com/elastic/cloud-on-k8s/test/e2e/test.RunMutations\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/run_mutation.go:39\ngithub.com/elastic/cloud-on-k8s/test/e2e/apm.TestAPMServerVersionUpgradeToLatest8x\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/test/e2e/apm/upgrade_test.go:33\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:1259"}
    --- FAIL: TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01 (300.00s)
```

I'm trying to understand why.

Reply:
@barkbay I'm working to replicate this failure and investigate.

Reply:
This test is still failing when run locally (`TestAPMServerVersionUpgradeToLatest8x` from `7.17.1` to `8.1.1`)
Looking at the APM Server logs I can see 2 security related errors at the beginning, then I _think_ the test is able to create the events, but it still eventually fails.

```
{"log.level":"error","@timestamp":"2022-04-28T08:03:44.099Z","log.logger":"modelindexer","log.origin":{"file.name":"modelindexer/indexer.go","file.line":395},"message":"failed to index event (security_exception): action [indices:admin/auto_create] is unauthorized for user [e2e-mercury-apmserver-upgrade-49wz-apm-user] with roles [apm_system,eck_apm_user_role_v75,ingest_admin] on indices [logs-apm.error-default], this action is granted by the index privileges [auto_configure,create_index,manage,all]","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2022-04-28T08:03:44.099Z","log.logger":"modelindexer","log.origin":{"file.name":"modelindexer/indexer.go","file.line":395},"message":"failed to index event (security_exception): action [indices:admin/auto_create] is unauthorized for user [e2e-mercury-apmserver-upgrade-49wz-apm-user] with roles [apm_system,eck_apm_user_role_v75,ingest_admin] on indices [metrics-apm.app.1234_service_12a3-default], this action is granted by the index privileges [auto_configure,create_index,manage,all]","service.name":"apm-server","ecs.version":"1.6.0"}
```

```
=== RUN   TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01
Retries (5m0s timeout): .
    step.go:43: 
        	Error Trace:	utils.go:88
        	Error:      	Received unexpected error:
        	            	timeout reached after 5m0s
        	Test:       	TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01
{"log.level":"error","@timestamp":"2022-04-28T10:08:40.998+0200","message":"stopping early","service.version":"0.0.0-SNAPSHOT+00000000","service.type":"eck","ecs.version":"1.4.0","error":"test failure","error.stack_trace":"github.com/elastic/cloud-on-k8s/test/e2e/test.StepList.RunSequential\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/step.go:44\ngithub.com/elastic/cloud-on-k8s/test/e2e/test.RunMutations\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/run_mutation.go:39\ngithub.com/elastic/cloud-on-k8s/test/e2e/apm.TestAPMServerVersionUpgradeToLatest8x\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/test/e2e/apm/upgrade_test.go:32\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:1439"}
    --- FAIL: TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01 (300.00s)
```

I'm still trying to understand if I'm doing something wrong...